docs/05-rfcs/2023-01-11-sdk-spec.md

```diff
@@ -0,0 +1,1043 @@
+   * Run an inflight whenever a file is created, uploaded, or deleted from the bucket.
+   */
+  on_event(fn: inflight (event: BucketEvent) => void, opts: BucketOnEventProps?): void;
+
+  /**
+   * Upload a file to the bucket.
+   */
+  inflight put(key: str, value: Serializable): void;
+
+  /**
+   * Get a file from the bucket.
+   */
+  inflight get(key: str): Serializable;
+
+  /**
+   * Delete a file from the bucket.
+   */
+  inflight delete(key: str): void;
+
+  /**
+   * List all files in the bucket with the given prefix.
+   */
+  inflight list(prefix: str?): Iterator<str>;
+
+  /**
+   * Returns a url to the given file.
+   * @throws Will throw if the public is not public.
+   */
+  inflight public_url(key: str): str;
+
+  /**
+   * Returns a signed url to the given file. This URL can be used by anyone to
+   * access the file until theh link expires (defaults to 24 hours).
+   */
+  inflight signed_url(key: str, duration?: duration): str;
+}
+```
+
+Future extensions:
+- `versioned` constructor property for object versioning
+
+## Queue
+
+> Note: this API is WIP, and needs more research for multi-cloud support.
+> Open questions:
+> - What is the API for FIFO queues?
+> - What is the API for dead-letter queues?
+> - What is the API for message visibility timeouts?
+> - What is the API for message retention periods?
+> - What is the API for message deduplication?
+> - Should the API use "push"/"pop" terminology or "send"/"receive" terminology?
+
+The queue resource represents a message buffer that can be used to decouple workloads between a set of producers and a set of consumers.
+Using a queue, you can send, store, and receive messages between software components at any volume, without losing messages or requiring other services to be available.
+
+Any number of producers can push messages to the queue, and any number of consumers can pop messages from the queue.
+When a consumer function receives a batch of messages, it is responsible for processing each message in the batch and acknowledging that the message has been processed by deleting it from the queue.
+If a consumer function does not acknowledge a message, the message will be re-delivered to another consumer function.
+
+A queue is not FIFO (first-in-first-out), so messages may be delivered out of order.
+
+```ts
+struct QueueProps {
+  /**
+   * How long a message is available for consumption before it's made available
+   * for other consumers.
+   */
+  timeout: duration?;
+}
+
+struct QueueAddConsumerProps { /* elided */ }
+
+resource Queue {
+  init(props: QueueProps = {});
+
+  /**
+   * Run an inflight in a cloud function whenever a message is pushed to the queue. 
+   */
+  add_consumer(fn: inflight (message: Serializable) => void, opts: QueueAddConsumerProps?): void;
+
+  /**
+   * Push a message to the queue.
+   */
+  inflight push(message: Serializable): void;
+  /**
+   * Pops (deletes and returns) a message from the queue.
+   */
+  inflight pop(): Serializable;
```

Comment:
would it be valuable to add an approximate message count? It looks like I can find a couple times on StackOverflow that devs have been looking for a good way to do this. 
[GCP](https://stackoverflow.com/questions/67329653/how-to-get-task-count-in-gcp-cloud-task)
[AWS](https://stackoverflow.com/questions/25351145/get-number-of-messages-in-an-amazon-sqs-queue)
[AZURE](https://stackoverflow.com/questions/11129761/unable-to-get-queue-length-message-count-from-azure)
```js
/**
* Retrieves approximate message count from queue
*/
inflight messageCount(): number
```

Reply:
Yes, totally :) what do you think of `approx_size`, or `size_approx` as an API name?

Reply:
I think approx_size would be my vote